packages/meditrak-server/src/apiV2/middleware/emailAfterTimeout.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { respond } from '@tupaia/utils';
+import { sendEmail } from '../../utilities';
+
+const sendResponseAsEmail = (user, subject, message) => {
+  const text = `Hi ${user.first_name},
+
+${message}
+  `;
+  sendEmail(user.email, subject, text);
+};
+
+const setupEmailResponse = async (req, res, constructEmailFromResponse) => {
+  const { models, userId } = req;
+  const user = await models.user.findById(userId);
+
+  if (res.headersSent) {
+    // no need to do anything if the endpoint handler responded successfully within the timeout
+    // n.b. this check needs to stay below any async stuff above, so that we don't end up in a half
+    // way state where the successful response happens e.g. during looking up the user
+    return;
+  }
+
+  // respond with timeout status
+  req.flagPermissionsChecked(); // any permissions error will be emailed; bypass permissions assertion
+  respond(res, { emailTimeoutHit: true });
+
+  // override the respond function so that when the endpoint handler finishes (or throws an error),
+  // the response is sent via email
+  res.overrideRespond = responseBody => {
+    const { subject, message } = constructEmailFromResponse(responseBody);
+    sendResponseAsEmail(user, subject, message);
+  };
+};
+
+// if the import takes too long, the results will be emailed
+export const emailAfterTimeout = constructEmailFromResponse => (req, res, next) => {
```

Comment:
Is there a reason that we are passing in the `constructEmailFromResponse` function as an argument here? Rather than just importing that function into this file and calling it directly?  ie.
```suggestion
export const emailAfterTimeout = (req, res, next) => {
```
I suppose it's because there may be other routes that want to use a different function for constructing the email? I mainly ask cos I got a bit confused trying to follow the code flow in this PR as I didn't realise that `constructEmailFromResponse` was a parameter instead of just the function itself.

Totally cool if we want it as a parameter if we anticipate different routes having different email building logic. I feel like at that point we'd want the original `constructEmailFromResponse` function to be given a more specific name, eg. `constructImportEmailFromResponse`, but keep this parameter the same?

Not really making any suggestion for change here, just flagging that I did get a little lost for what it's worth 👍

Reply:
Yes that's exactly right! In a coming change this will be applied across both import and export routes, which each have their own requirements for interpreting the response that would normally be sent to the front end, and converting it into an email. Good suggestion for naming the original function something more specific - I've done that for the export one but will change the import one to do the same in the PR I'm working on atm. Will leave it named as is here to avoid the upstream merge chain :-)

I've tried to make this series of PRs quite granular, but sometimes that makes it harder for the reviewer to see the bigger picture. 

Interested in your perspective there.... 

It's working well on my end, as it encourages me to focus things into discrete tasks (and Linear's subtasks being super lightweight help with that too), which I don't traditionally do. That's pretty helpful because I spend so much of my day distracted by whatever meetings or other things are going on that having a contained thing to come back to feels like I can be more productive in the non-distracted hours :-)